File: backend/app/zero_trust/zero_trust_api.py
from typing import List, Dict, Any
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.database.connection import get_db
from app.zero_trust.zero_trust_service import ZeroTrustService

router = APIRouter(tags=["Zero Trust & Attack Simulator (Module 6)"])


class ZeroTrustVerificationRequest(BaseModel):
    session_uuid: str
    packet_id: str
    source_node: str
    command: str
    hmac_signature: str
    nonce: str
    sequence_number: int


class ZeroTrustVerificationResponse(BaseModel):
    decision: str
    trust_score: float
    risk_level: str
    checks_passed: int
    total_checks: int = 20
    failed_checks: List[str]
    rationale: Dict[str, Any]
    timestamp: str


class AttackSimulationRequest(BaseModel):
    session_uuid: str
    attack_type: str
    intensity: float = 1.0


class AttackSimulationResponse(BaseModel):
    attack_id: str
    session_uuid: str
    attack_type: str
    detected: bool
    mitigation_action: str
    trust_score_impact: float
    details: str
    timestamp: str


@router.post("/zero-trust/verify", response_model=ZeroTrustVerificationResponse)
async def verify_zero_trust(
    request: ZeroTrustVerificationRequest,
    db: AsyncSession = Depends(get_db)
) -> ZeroTrustVerificationResponse:
    """Executes 20-stage zero-trust security pipeline on a SCADA command packet."""
    service = ZeroTrustService(db)
    log = await service.verify_packet(
        session_uuid=request.session_uuid,
        packet_id=request.packet_id,
        source_node=request.source_node,
        command=request.command,
        hmac_signature=request.hmac_signature,
        nonce=request.nonce,
        sequence_number=request.sequence_number
    )
    return ZeroTrustVerificationResponse(
        decision=log.decision,
        trust_score=log.trust_score,
        risk_level=log.risk_level,
        checks_passed=log.checks_passed,
        total_checks=20,
        failed_checks=log.failed_checks or [],
        rationale=log.rationale or {},
        timestamp=log.timestamp.isoformat()
    )


@router.post("/attacks/simulate", response_model=AttackSimulationResponse)
async def simulate_attack(
    request: AttackSimulationRequest,
    db: AsyncSession = Depends(get_db)
) -> AttackSimulationResponse:
    """Simulates quantum or classical network attacks to test zero-trust defenses."""
    service = ZeroTrustService(db)
    res = await service.simulate_attack(
        session_uuid=request.session_uuid,
        attack_type=request.attack_type,
        intensity=request.intensity
    )
    return AttackSimulationResponse(**res)

File: frontend/src/services/zeroTrustApi.ts
import { apiRequest } from '../core/apiClient';

export interface ZeroTrustVerificationRequest {
  session_uuid: string;
  packet_id: string;
  source_node: string;
  command: string;
  hmac_signature: string;
  nonce: string;
  sequence_number: number;
}

export interface ZeroTrustVerificationResponse {
  decision: 'ALLOW' | 'BLOCK';
  trust_score: number;
  risk_level: string;
  checks_passed: number;
  total_checks: number;
  failed_checks: string[];
  rationale: Record<string, any>;
  timestamp: string;
}

export interface AttackSimulationRequest {
  session_uuid: string;
  attack_type: 'EAVESDROPPING' | 'MAN_IN_THE_MIDDLE' | 'REPLAY_ATTACK' | 'PACKET_TAMPERING' | 'DENIAL_OF_SERVICE';
  intensity: number;
}

export interface AttackSimulationResponse {
  attack_id: string;
  session_uuid: string;
  attack_type: string;
  detected: boolean;
  mitigation_action: string;
  trust_score_impact: number;
  details: string;
  timestamp: string;
}

export async function verifyZeroTrust(payload: ZeroTrustVerificationRequest): Promise<ZeroTrustVerificationResponse> {
  return apiRequest<ZeroTrustVerificationResponse>('/zero-trust/verify', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function simulateAttack(payload: AttackSimulationRequest): Promise<AttackSimulationResponse> {
  return apiRequest<AttackSimulationResponse>('/attacks/simulate', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}
